### PR TITLE
✨ Add support for multi-type data requests

### PR DESCRIPTION
**Pull request by burkematthew**
This PR addresses #256 and #344 .

@jaron-l @mrooney here is the draft PR to address the feature request.  I have more work to do, but wanted to at least provide some visibility to how I am going about this implementation.

### CODE REVIEW

**Comment by burkematthew** [comment]
@jaron-l any thoughts on the approach before I continue?

**Review by jaron-l — COMMENTED** [review]
I like the effort that you put into this and I think it might be a good option, but it kinda seems like it might be better to push for a breaking change here. To understand the code, you really have to walk through it which means it's complicated. It might be cleaner in this scenario to just go for a breaking change and then write the output data really cleanly with keys that always describe the data that is coming out. @mrooney any thoughts on this?

File: mintapi/api.py [hunk 1]
```diff
@@ -1767,78 +1811,63 @@ def main():
...
+        print(json.dumps(data, indent=2))
+    elif options.filename.endswith(".json"):
+        with open(options.filename, "w+") as f:
+            json.dump(data, f, indent=2)
+    elif (
+        (options.transactions or options.extended_transactions)
+        and overlapping_options == 1
+        and options.filename.endswith(".csv")
+    ):
+        print(pd.json_normalize(data))
```
[the remark this answers is not in the document]
Comment (burkematthew) [reply mintapi/api.py:1868]: @jaron-l yes, it's just not complete yet. I submitted the state of what I had so far.

**Comment by burkematthew** [comment]
> I like the effort that you put into this and I think it might be a good option, but it kinda seems like it might be better to push for a breaking change here. To understand the code, you really have to walk through it which means it's complicated. It might be cleaner in this scenario to just go for a breaking change and then write the output data really cleanly with keys that always describe the data that is coming out. @mrooney any thoughts on this?

@jaron-l @mrooney 

I'm not 100% sure I understand what would be the breaking change.  In some ways, it seems like your suggestion to "write the output data really cleanly with keys that always describe the data" comment is a different enhancement than allowing the return of multiple types of data at once.

In one sense, I think the idea behind this enhancement is to create one large JSON Response object from `mintapi`, the types of data included are configurable by the enduser.  Then, our CSV functionality is really just to create a CSV-formatted output for each key within that main object, i.e. `accounts`, `transactions`, `budgets`, etc.  I think this latter functionality would be more of a breaking change as we'd need to create multiple CSVs and so that would be fundamentally different.  However, I think at least allowing transactions to include with accounts and budgets (in JSON format) can be done in a non-breaking manner, while meeting requests made by several different users and staying in line with a similar pattern that we already have in place.

**Comment by mrooney** [comment]
Yeah, I guess there are a couple options for larger changes. One would be to grab a ton of data and output it all and let you grab the keys you need, and another would be something like GraphQL where the user specifies the format and we grab what we need and match it.

However, to address those feature requests I don't think we need anything crazy. Just allowing multiple lines in the config-file option and iterating over them would be fine, and only logging in once, right?

**Comment by burkematthew** [comment]
@mrooney yes that is correct. You can see the attempt made here which does basically that. 

I do like the idea for an enhancement down the road to look at GraphQL. That could be interesting.

**Comment by jaron-l** [comment]
@burkematthew, I think they should be considered together because to continue down this path would be to make the code continue to get more complicated, and I also think, make the output more complicated and harder to work with. I show what I mean.

The output right now is complicated because you have to know how the code works in order to understand the output. For example, if I call for budget data, I get the following as output:

```json
{
    budget_data
}
```

but if I call for budgets and account data, I get the following:

```json
{
     "budgets" : budget_data,
     "accounts": account_data,
}
```

If you don't know the code, one might ask, why did the budgets_data value move one layer down in the hierarchy?

Essentially what your code is doing is checking whether the hierarchy was two layers or one layer before and trying to maintain backwards compatibility to make the output the same as before, but if multiple types are found, go ahead and break the hierarchy and add a level to the json output. I'm just saying, let's be consistent and always have the same depth to the output so that different combinations of flags don't produce different levels of Json output. For example:

calling budgets would produce:

```json
{
     "budgets": budgets_data
}
```

 and calling budgets and accounts would produce:

```json
{
    "budgets": budgets_data,
    "accounts":, accounts_data,
}
```
which is a breaking change but doesn't require users to read data at different levels of output depending on what they decided to call mintapi with.

**Comment by burkematthew** [comment]
@jaron-l while I understand what you're trying to say, I disagree.  The whole advantage of JSON format and key\data pairing formats is that the order of the objects within the Response does not matter.  Hierarchy of JSON is a parent\child relationship, not a sibling relationship.  So, while a user might not understand why the order changed within the Response, it does actually change the value of the object and so wouldn't be considered a breaking change.

I'd also argue that someone that specifically requests multiple types of data (in JSON format) would expect an object with two keys of, for example, `budgets` and `accounts`.

**Comment by jaron-l** [comment]
> @jaron-l while I understand what you're trying to say, I disagree. The whole advantage of JSON format and key\data pairing formats is that the order of the objects within the Response does not matter. Hierarchy of JSON is a parent\child relationship, not a sibling relationship. So, while a user might not understand why the order changed within the Response, it does actually change the value of the object and so wouldn't be considered a breaking change.

I think my argument has more to do with how it simplifies code on our side and the user's side rather than the ability of JSON to not care about formatting. For example, instead of this code:

```python
def _process_final_data(key, data, new_data, overlapping_options):
    if overlapping_options > 1:
        # Key assignment cannot be done when data is None; so we initialize if it is None.
        if data is None:
            data = {}
        data[key] = new_data
    else:
        data = new_data
    return data


def _process_final_transaction_data(data, transactions, overlapping_options):
    return _process_final_data(
        "transactions",
        data,
        transactions.to_json(orient="records"),
        overlapping_options,
    )
overlapping_options = sum(
        (key == "budgets" and value == True)
        or ((key == "transactions" or key == "extended_transactions") and value == True)
        or (key == "accounts" and value == True)
        for key, value in options.__dict__.items()
    )
if options.extended_transactions:
        transactions = mint._get_detailed_transactions(options)
        # Because transactions is a Pandas Data Frame, we have to check for empty.
        if not transactions.empty:
            data = _process_final_transaction_data(
                data, transactions, overlapping_options
            )
```
We could simply have this:
```python
if options.extended_transactions:
    data["transactions"] = mint.get_transactions(...)
```
and that same simple structure could be for all command line options without checking for nesting.

On the users side, the structure would be more scalable. If they were previously calling just one flag and switched to adding more commands, their previous parsing function would still be valid and would require no modifications. In our current setup, they would have to change their code and the exact structure would be undefined unless they ran it first as it depends on what they called before to what changes will happen.
 
> I'd also argue that someone that specifically requests multiple types of data (in JSON format) would expect an object with two keys of, for example, `budgets` and `accounts`, so the sibling positioning would look a little different.

That's fair but why not make it even more intuitive by not having a change at all?

> Also, I think we could probably sort the individual keys however we want to.

I'm not sure what you mean here. Are we talking about the order of the keys on the first level of the JSON object? Why does the order of the keys come into this conversation?

To have a little more context however, my suggestion would of necessity have to wait until v2 anyways. I'm more arguing that it's not worth complicating the code and maybe push to have v2 sooner rather than fix this and remove it later.

**Pull request closed without merging**